types/module.go

```diff
@@ -0,0 +1,233 @@
+	AppModuleBasic
+
+	// registers
+	RegisterInvariants(InvariantRouter)
+
+	// routes
+	Route() string
+	NewHandler() Handler
+	QuerierRoute() string
+	NewQuerierHandler() Querier
+
+	// genesis
+	InitGenesis(Context, json.RawMessage) []abci.ValidatorUpdate
+	ExportGenesis(Context) json.RawMessage
+
+	BeginBlock(Context, abci.RequestBeginBlock) Tags
+	EndBlock(Context, abci.RequestEndBlock) ([]abci.ValidatorUpdate, Tags)
+}
+
+// module manager provides the high level utility for managing and executing
+// operations for a group of modules
+type ModuleManager struct {
+	Modules            map[string]AppModule
+	OrderInitGenesis   []string
+	OrderExportGenesis []string
+	OrderBeginBlockers []string
+	OrderEndBlockers   []string
+}
+
+// NewModuleManager creates a new ModuleManager object
+func NewModuleManager(modules ...AppModule) *ModuleManager {
+
+	moduleMap := make(map[string]AppModule)
+	var modulesStr []string
+	for _, module := range modules {
+		moduleMap[module.Name()] = module
+		modulesStr = append(modulesStr, module.Name())
+	}
+
+	return &ModuleManager{
+		Modules:            moduleMap,
+		OrderInitGenesis:   modulesStr,
+		OrderExportGenesis: modulesStr,
+		OrderBeginBlockers: modulesStr,
+		OrderEndBlockers:   modulesStr,
+	}
+}
+
+// set the order of init genesis calls
+func (mm *ModuleManager) SetOrderInitGenesis(moduleNames ...string) {
+	mm.OrderInitGenesis = moduleNames
+}
+
+// set the order of export genesis calls
+func (mm *ModuleManager) SetOrderExportGenesis(moduleNames ...string) {
+	mm.OrderExportGenesis = moduleNames
+}
+
+// set the order of set begin-blocker calls
+func (mm *ModuleManager) SetOrderBeginBlockers(moduleNames ...string) {
+	mm.OrderBeginBlockers = moduleNames
+}
+
+// set the order of set end-blocker calls
+func (mm *ModuleManager) SetOrderEndBlockers(moduleNames ...string) {
+	mm.OrderEndBlockers = moduleNames
+}
+
+// register all module routes and module querier routes
+func (mm *ModuleManager) RegisterInvariants(invarRouter InvariantRouter) {
+	for _, module := range mm.Modules {
+		module.RegisterInvariants(invarRouter)
+	}
+}
+
+// register all module routes and module querier routes
+func (mm *ModuleManager) RegisterRoutes(router Router, queryRouter QueryRouter) {
+	for _, module := range mm.Modules {
+		if module.Route() != "" {
+			router.AddRoute(module.Route(), module.NewHandler())
+		}
+		if module.QuerierRoute() != "" {
+			queryRouter.AddRoute(module.QuerierRoute(), module.NewQuerierHandler())
+		}
+	}
+}
+
+// perform init genesis functionality for modules
+func (mm *ModuleManager) InitGenesis(ctx Context, genesisData map[string]json.RawMessage) abci.ResponseInitChain {
+	var validatorUpdates []abci.ValidatorUpdate
+	for _, moduleName := range mm.OrderInitGenesis {
+		if genesisData[moduleName] == nil {
+			continue
+		}
+		moduleValUpdates := mm.Modules[moduleName].InitGenesis(ctx, genesisData[moduleName])
+
+		// use these validator updates if provided, the module manager assumes
```

Comment:
If we do this I think we should explicitly panic if more than one module returns validator set updates.

Reply:
k here, there are updates from both staking and genesis-tx stuff.. I wonder should we really be appending them? currently genutil (called last) is overwriting any updates from staking, is this a bug do you think?

Reply:
++

Reply:
yeah so what's happening here is consistent with master (https://github.com/cosmos/cosmos-sdk/blob/master/cmd/gaia/app/app.go#L281-L315) sounds like something to get to the bottom of... I imagine we should be appending them.... thoughts anyone?

Reply:
Basically, Staking Initchain produces abci validator updates, but GenTx logic _also_ does. Currently GenTx updates will just overwrite any staking updates if they’re both present

I guess this hasn’t been an issue because when we dump genesis and restart the chain we don’t include GenTx’s - and then we use GenTxs we don’t have any validators
hmm interesting
So it really _is_ okay to just use one or the other… I guess we just need to include this logic in ValidateGenesis as well as a panic in InitChain

Reply:
So what is the resolution here? I think it's super confusing to have validator updates comes from varying sources. What is the ideal/preferred canonical source of truth? InitChain or GenTxs?

Reply:
Currently we need to allow for both because there are multiple modes which the genesis can be used in (which we should eventually make explicit see: https://github.com/cosmos/cosmos-sdk/issues/2862) - I don't think it's a problem for the interim, we just need good code comments as well as a panic checks and also an explicit check for this in `ValidateGenesis`, I'm going to fix this right now and I'll reference the commit here

Reply:
So turns out that the best way to check in ValidateGenesis is to just make sure that the init chain doesn't panic - however this is a bit of a struggle so I'm just replacing with a TODO for now. I think that we should fully address this in `ValidateGenesis` as a part of #2862